packager/media/base/aes_decryptor.cc

```diff
@@ -112,44 +115,49 @@ bool AesCbcDecryptor::CryptInternal(const uint8_t* ciphertext,
-  uint8_t* next_to_last_plaintext_block =
-      plaintext + ciphertext_size - residual_block_size - AES_BLOCK_SIZE;
-
-  // Determine what the last IV should be so that we can "skip ahead" in the
-  // CBC decryption.
-  std::vector<uint8_t> last_iv(
-      ciphertext + ciphertext_size - residual_block_size,
-      ciphertext + ciphertext_size);
-  last_iv.resize(AES_BLOCK_SIZE, 0);
-
-  // Decrypt the next-to-last block using the IV determined above. This decrypts
-  // the residual block bits.
-  AES_cbc_encrypt(next_to_last_ciphertext_block, next_to_last_plaintext_block,
-                  AES_BLOCK_SIZE, aes_key(), last_iv.data(), AES_DECRYPT);
-
-  // Swap back the residual block bits and the next-to-last block.
-  if (plaintext == ciphertext) {
-    std::swap_ranges(next_to_last_plaintext_block,
-                     next_to_last_plaintext_block + residual_block_size,
-                     next_to_last_plaintext_block + AES_BLOCK_SIZE);
-  } else {
-    memcpy(next_to_last_plaintext_block + AES_BLOCK_SIZE,
-           next_to_last_plaintext_block, residual_block_size);
-    memcpy(next_to_last_plaintext_block,
-           next_to_last_ciphertext_block + AES_BLOCK_SIZE, residual_block_size);
+    CbcDecryptBlocks(ciphertext, cbc_size - AES_BLOCK_SIZE, plaintext);
   }
 
-  // Decrypt the next-to-last full block.
-  AES_cbc_encrypt(next_to_last_plaintext_block, next_to_last_plaintext_block,
-                  AES_BLOCK_SIZE, aes_key(), internal_iv_.data(), AES_DECRYPT);
+  uint8_t* next_to_last_plaintext_block = plaintext + cbc_size - AES_BLOCK_SIZE;
+
+  // The next-to-last block should be decrypted first in ECB mode, which is
+  // effectively what you get with an IV of all zeroes.
+  std::vector<uint8_t> backup_iv(internal_iv_);
+  internal_iv_.assign(AES_BLOCK_SIZE, 0);
+  // mbedtls requires a buffer large enough for one extra block.
+  std::vector<uint8_t> stolen_bits(AES_BLOCK_SIZE * 2);
+  CbcDecryptBlocks(next_to_last_block.data(), AES_BLOCK_SIZE,
+                   stolen_bits.data());
+
+  // Reconstruct the final two blocks of ciphertext.
+  std::vector<uint8_t> reconstructed_blocks(AES_BLOCK_SIZE * 2);
+  memcpy(reconstructed_blocks.data(), residual_block.data(),
+         residual_block_size);
+  memcpy(reconstructed_blocks.data() + residual_block_size,
+         stolen_bits.data() + residual_block_size,
+         AES_BLOCK_SIZE - residual_block_size);
+  memcpy(reconstructed_blocks.data() + AES_BLOCK_SIZE,
+         next_to_last_block.data(), AES_BLOCK_SIZE);
+
+  // Decrypt the last two blocks.
+  internal_iv_ = backup_iv;
+  // mbedtls requires a buffer large enough for one extra block.
+  std::vector<uint8_t> final_output_blocks(AES_BLOCK_SIZE * 3);
+  CbcDecryptBlocks(reconstructed_blocks.data(), AES_BLOCK_SIZE * 2,
+                   final_output_blocks.data());
+
+  // Copy the final output.
+  memcpy(next_to_last_plaintext_block, final_output_blocks.data(),
+         AES_BLOCK_SIZE + residual_block_size);
```

Comment:
Just two nits (feel free to ignore):

1. The original implementation decrypts one less block: the new implementation decrypts the residual block twice while the old implementation decrypts the residual block only once.
2. CTS is only needed for Widevine classic packaging, which is already deprecated.

Reply:
I understand.  I couldn't make the original implementation work, though, possibly due to mbedtls treating overlapping input/output buffers differently, or the way it treats IV updates differently.

I never did figure out what was wrong with my attempt to port the old implementation more directly, but this re-implementation from descriptions of CTS in Wikipedia allowed me to create this, which passes all the original tests.  I'm inclined to keep it for now, though, at least until we choose to separately remove Widevine Classic support.  (It's still in there, right?)

Reply:
Yah, it is a nit. Feel free to ignore.

> (It's still in there, right?)

IIRC, there are still a few long tail use cases, but we don't expect new contents to be packaged. The plan is to fully deprecate it by the end of 2022 or in early 2023.

Reply:
OK, sounds good.  Thanks!